cmd/snap-confine/mount-support.c

```diff
@@ -679,37 +679,4 @@ void sc_populate_mount_ns(const char *base_snap_name, const char *snap_name)
 	}
 }
 
-static bool is_mounted_with_shared_option(const char *dir)
-    __attribute__ ((nonnull(1)));
 
-static bool is_mounted_with_shared_option(const char *dir)
-{
-	struct sc_mountinfo *sm SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
-	sm = sc_parse_mountinfo(NULL);
-	if (sm == NULL) {
-		die("cannot parse /proc/self/mountinfo");
-	}
-	struct sc_mountinfo_entry *entry = sc_first_mountinfo_entry(sm);
-	while (entry != NULL) {
-		const char *mount_dir = entry->mount_dir;
-		if (sc_streq(mount_dir, dir)) {
-			const char *optional_fields = entry->optional_fields;
-			if (strstr(optional_fields, "shared:") != NULL) {
-				return true;
-			}
-		}
-		entry = sc_next_mountinfo_entry(entry);
-	}
-	return false;
-}
-
-void sc_ensure_shared_snap_mount(void)
-{
-	if (!is_mounted_with_shared_option("/")
-	    && !is_mounted_with_shared_option(SNAP_MOUNT_DIR)) {
-		sc_do_mount(SNAP_MOUNT_DIR, SNAP_MOUNT_DIR, "none",
```

Comment:
We could actually keep this code but only change it to do the sharing, not the bind mounting.

The bind mounting was the problem because the previously mounted snaps are still mounted, just under the bind mount layer. With the new mount unit perhaps all we need is `sc_do_mount("none", SNAP_MOUNT_DIR, NULL, MS_SHARED | MS_REC, NULL);` here.